In [36]:
import pandas as pd

# Initialize DataFrame
df = pd.read_csv('BCGenAI_Task1.csv')

In [37]:
# Convert the cleaned string columns to numeric (Total Revenue + Net Income)
df['Total Revenue (in millions)'] = pd.to_numeric(df['Total Revenue (in millions)'], errors='coerce')
df['Net Income (in millions)'] = pd.to_numeric(df['Net Income (in millions)'], errors='coerce')

# Sort by Company and Fiscal Year
df.sort_values(by=['Company', 'Fiscal Year'], inplace=True)

# Calculate percentage changes
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue (in millions)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income (in millions)'].pct_change() * 100

# Mean Revenue and Net Income, by Company
df['Mean Revenue (in millions)'] = df.groupby(['Company'])['Total Revenue (in millions)'].transform('mean')
df['Mean Net Income (in millions)'] = df.groupby(['Company'])['Net Income (in millions)'].transform('mean')

# Display the DataFrame with manual verification columns - Revenue and Net Income (Growth + Mean)
df[['Company', 'Fiscal Year', 'Total Revenue (in millions)', 'Revenue Growth (%)', 'Mean Revenue (in millions)', 
    'Net Income (in millions)', 'Net Income Growth (%)', 'Mean Net Income (in millions)']]

Company  Fiscal Year  Total Revenue (in millions)  Revenue Growth (%)  \
8      Apple         2021                       365817                 NaN   
7      Apple         2022                       394328            7.793788   
6      Apple         2023                       383285           -2.800461   
2  Microsoft         2021                       168088                 NaN   
1  Microsoft         2022                       198270           17.956071   
0  Microsoft         2023                       211915            6.882030   
5      Tesla         2021                        53823                 NaN   
4      Tesla         2022                        81462           51.351653   
3      Tesla         2023                        96773           18.795267   

   Mean Revenue (in millions)  Net Income (in millions)  \
8               381143.333333                     94680   
7               381143.333333                     99803   
6               381143.333333                     96995   
2               192757.666667                     61271   
1               192757.666667                     72738   
0               192757.666667                     72361   
5                77352.666667                      5644   
4                77352.666667                     12587   
3                77352.666667                     14974   

   Net Income Growth (%)  Mean Net Income (in millions)  
8                    NaN                   97159.333333  
7               5.410858                   97159.333333  
6              -2.813543                   97159.333333  
2                    NaN                   68790.000000  
1              18.715216                   68790.000000  
0              -0.518299                   68790.000000  
5                    NaN                   11068.333333  
4             123.015592                   11068.333333  
3              18.964010                   11068.333333

## Analyzing Total Revenue + Net Income Trends

As shown by the derived DataFrame above, the highest growth percentage for both Revenue and Net Income was by Tesla between FY2021 and FY2022, with values of (51.35%) and (123.02%) respectively. The next highest was also by Tesla, between FY2022 and FY2023, with values of (18.8%) and (18.96%) respectively. Overall, Microsoft had the next highest growth in both Revenue and Net Income, while Apple had the lowest overall Revenue and Net Income growth in all fiscal years. However, it is important to note that while Apple did have the least growth, it stayed consistent in the highest total revenue and net income values for all three fiscal years.

In terms of growth patterns between fiscal years, all three companies proved a lower overall growth between FY2022 and FY2023 in comparison to their prior fiscal year comparison. As can be seen in our analysis, both Apple and Microsoft reflected negative growth in this years, with negative net income growth (-0.52%) for Microsoft, and negative growth in both revenue (-2.8%) and net income (-2.81%) for Apple.

Finally, for our derived mean values, Apple has the highest mean revenue (381,143.33M) and net income (79,159.33M) among the fiscal years, as discussed previously. Between Microsoft and Tesla, Microsoft had a much higher mean revenue (192,757.67M) and mean net income (68,790M) than Tesla, which had mean values of (77,352.67M) and (11,086.33M) respectively. 

In [38]:
# Convert the cleaned string columns to numeric (Total Assets + Liabilities + Cash Flow (Operating))
df['Total Assets (in millions)'] = pd.to_numeric(df['Total Assets (in millions)'], errors='coerce')
df['Total Liabilities (in millions)'] = pd.to_numeric(df['Total Liabilities (in millions)'], errors='coerce')
df['Cash Flow from Operating Activities (in millions)'] = pd.to_numeric(df['Cash Flow from Operating Activities (in millions)'], errors='coerce')

# Sort by Company and Fiscal Year
df.sort_values(by=['Company', 'Fiscal Year'], inplace=True)

# Calculate percentage changes
df['Asset Growth (%)'] = df.groupby(['Company'])['Total Assets (in millions)'].pct_change() * 100
df['Liability Growth (%)'] = df.groupby(['Company'])['Total Liabilities (in millions)'].pct_change() * 100
df['Cash Flow Growth - Operating (%)'] = df.groupby(['Company'])['Cash Flow from Operating Activities (in millions)'].pct_change() * 100

# Mean Assets, Liabilities, and Cash Flow, by Company
df['Mean Assets (in millions)'] = df.groupby(['Company'])['Total Assets (in millions)'].transform('mean')
df['Mean Liabilities (in millions)'] = df.groupby(['Company'])['Total Liabilities (in millions)'].transform('mean')
df['Mean Cash Flow - Operating (in millions)'] = df.groupby(['Company'])['Cash Flow from Operating Activities (in millions)'].transform('mean')

# Display the DataFrame with manual verification columns - Assets, Liabilities, and Cash Flow (Growth + Mean)
df[['Company', 'Fiscal Year', 'Total Assets (in millions)', 'Asset Growth (%)', 'Mean Assets (in millions)', 
    'Total Liabilities (in millions)', 'Liability Growth (%)', 'Mean Liabilities (in millions)', 'Cash Flow from Operating Activities (in millions)',
       'Cash Flow Growth - Operating (%)', 'Mean Cash Flow - Operating (in millions)']]

Company  Fiscal Year  Total Assets (in millions)  Asset Growth (%)  \
8      Apple         2021                      351002               NaN   
7      Apple         2022                      352755          0.499427   
6      Apple         2023                      352583         -0.048759   
2  Microsoft         2021                      333779               NaN   
1  Microsoft         2022                      364840          9.305858   
0  Microsoft         2023                      411976         12.919636   
5      Tesla         2021                       62131               NaN   
4      Tesla         2022                       82338         32.523217   
3      Tesla         2023                      106618         29.488207   

   Mean Assets (in millions)  Total Liabilities (in millions)  \
8              352113.333333                           287912   
7              352113.333333                           302083   
6              352113.333333                           290437   
2              370198.333333                           191791   
1              370198.333333                           198298   
0              370198.333333                           205753   
5               83695.666667                            30548   
4               83695.666667                            36440   
3               83695.666667                            43009   

   Liability Growth (%)  Mean Liabilities (in millions)  \
8                   NaN                   293477.333333   
7              4.921990                   293477.333333   
6             -3.855232                   293477.333333   
2                   NaN                   198614.000000   
1              3.392756                   198614.000000   
0              3.759493                   198614.000000   
5                   NaN                    36665.666667   
4             19.287678                    36665.666667   
3             18.026894                    36665.666667   

   Cash Flow from Operating Activities (in millions)  \
8                                             104038   
7                                             122151   
6                                             110543   
2                                              76740   
1                                              89035   
0                                              87582   
5                                              11497   
4                                              14724   
3                                              13256   

   Cash Flow Growth - Operating (%)  Mean Cash Flow - Operating (in millions)  
8                               NaN                             112244.000000  
7                         17.409985                             112244.000000  
6                         -9.502992                             112244.000000  
2                               NaN                              84452.333333  
1                         16.021631                              84452.333333  
0                         -1.631942                              84452.333333  
5                               NaN                              13159.000000  
4                         28.068192                              13159.000000  
3                         -9.970117                              13159.000000

## Analyzing Total Assets, Total Liabilities, and Cash Flow (Operating) Trends

Similar to the previous analysis, the highest growth percentage for Assets, Liabilities, and Cash Flow (Operating) was by Tesla between FY2021 and FY2022, with values of (32.52%), (19.29%), and (28.07%) respectively. However, although the next highest growth in Assets and Liabilities was also by Tesla between FY2022 and FY2023, with values of (29.49%) and (18.03), this was not the case for the Cash Flow Growth (Operating), which actually had the most negative growth across all companies and fiscal years (-9.97%). Instead, the highest Cash Flow Growth (Operating) was by Apple between FY2021 and FY2022 (17.41%). This can be considered an anomaly, as the rest of Apple's growth values in all fiscal years were the lowest across all companies. In fact, we can see that between FY2022 and FY2023, all of Apple's growth percentages for Assets (-0.05%), Liabilities (-3.86%), and Cash Flow (-9.5%) were negative.

Contrary to the previous analysis, in terms of growth patterns between fiscal years, the lowest growth rates were not all constricted to one year. For example, Microsoft had a higher Asset growth (12.92%) and Liability growth (3.76%) between FY2022 and FY2023, but it had a lower Cash Flow Growth for that period, with a negative growth of (-1.63%) vs. its positive growth of (16.02%) between FY2021 and FY2022. We can also conversely see that instead of a higher Asset growth like Microsoft, both Tesla and Apple had a lower Asset growth between FY2022 and FY2023 than FY2021 and FY2022. 

Finally, the derived means were also varied among all figures. Microsoft had the highest mean assets (379,198.33M), but Apple had the highest mean liabilites (293,477.33M) and operating cash flow (112,244M).

In [39]:
# Display the DataFrame with only the new calculations
df[['Company', 'Fiscal Year', 'Asset Growth (%)', 'Mean Assets (in millions)', 'Liability Growth (%)', 
    'Mean Liabilities (in millions)',
       'Cash Flow Growth - Operating (%)', 'Mean Cash Flow - Operating (in millions)']]

Company  Fiscal Year  Asset Growth (%)  Mean Assets (in millions)  \
8      Apple         2021               NaN              352113.333333   
7      Apple         2022          0.499427              352113.333333   
6      Apple         2023         -0.048759              352113.333333   
2  Microsoft         2021               NaN              370198.333333   
1  Microsoft         2022          9.305858              370198.333333   
0  Microsoft         2023         12.919636              370198.333333   
5      Tesla         2021               NaN               83695.666667   
4      Tesla         2022         32.523217               83695.666667   
3      Tesla         2023         29.488207               83695.666667   

   Liability Growth (%)  Mean Liabilities (in millions)  \
8                   NaN                   293477.333333   
7              4.921990                   293477.333333   
6             -3.855232                   293477.333333   
2                   NaN                   198614.000000   
1              3.392756                   198614.000000   
0              3.759493                   198614.000000   
5                   NaN                    36665.666667   
4             19.287678                    36665.666667   
3             18.026894                    36665.666667   

   Cash Flow Growth - Operating (%)  Mean Cash Flow - Operating (in millions)  
8                               NaN                             112244.000000  
7                         17.409985                             112244.000000  
6                         -9.502992                             112244.000000  
2                               NaN                              84452.333333  
1                         16.021631                              84452.333333  
0                         -1.631942                              84452.333333  
5                               NaN                              13159.000000  
4                         28.068192                              13159.000000  
3                         -9.970117                              13159.000000

## Final Conclusions

Above is a summarized charts of solely all the calculations we conducted in this notebook. Based off of these numbers and the trends we verbalized earlier, one clear standout was the observed variations in total assets, liabilities, and cash flow from operating activities compared to total revenue and net income.

Several possibilities can explain this:

#### Revenue and Net Income Trends:
Revenue and net income trends showed more consistent growth patterns, particularly for Tesla, which experienced the highest growth rates. This can be attributed to Tesla's aggressive market expansion, increased production capacity, and growing demand for electric vehicles. Conversely, Apple's relatively stable yet high revenue and net income values reflect its established market dominance and mature product lines, which typically see incremental growth rather than rapid increases.

#### Variability in Assets, Liabilities, and Cash Flow
The greater variation in total assets, liabilities, and cash flow from operating activities across fiscal years suggests several underlying factors:

#### Investment Strategies:
Companies like Tesla, which are in high-growth phases, often see significant fluctuations in assets and liabilities due to ongoing investments in new factories, technologies, and market expansion efforts. These investments can lead to substantial increases in both assets and liabilities as the company finances its growth through both equity and debt.

#### Market Conditions:
Fluctuations in cash flow from operating activities can be influenced by broader economic conditions, changes in consumer demand, and operational efficiencies. For instance, negative cash flow growth in certain years might reflect periods of heavy investment or downturns in market demand that impact operational revenues.

#### Operational Changes:
Variations in cash flow growth could also result from changes in operational efficiency, cost management, and production scaling. Companies like Microsoft, which have diversified product lines and services, might experience variability in cash flow due to shifts in operational focus or the introduction of new products and services.

#### Financial Management:
The differences in mean values and growth rates of liabilities and assets also highlight distinct financial management strategies. Apple, for instance, manages to maintain high cash flow despite lower growth rates, indicating effective cost control and a strong cash reserve strategy. In contrast, Tesla's higher variability reflects its ongoing investment in scaling operations and infrastructure.


### Broader Implications
Overall, the broader trends observed in the financial data suggest that while revenue and net income provide a snapshot of a company's market performance and profitability, a deeper analysis of assets, liabilities, and cash flow is crucial to understanding the underlying financial health and strategic priorities of the company. The variations in these metrics highlight the dynamic nature of financial management and the impact of strategic decisions on a company's financial trajectory.

*All code, analysis, and conclusions are developed by Naisha Sinha for the BCG GenAI Engineer Simulation.*